File: praja/backend/app/routes/sentinel.py
from datetime import datetime, timezone, timedelta
from fastapi import APIRouter, Depends
from typing import Any
from collections import defaultdict
from app.db.database import get_supabase
from app.utils.jwt import get_current_user

router = APIRouter()


@router.get("/summary")
def sentinel_summary(
    current: dict = Depends(get_current_user),
    sb: Any = Depends(get_supabase),
):
    scores = sb.table("ward_sentiment_scores").select("*").order("score", desc=True).limit(10).execute()
    grievances = sb.table("grievances").select("ai_category").eq("status", "open").execute()
    category_counts: dict[str, int] = {}
    for g in grievances.data:
        cat = g.get("ai_category", "General")
        category_counts[cat] = category_counts.get(cat, 0) + 1
    return {
        "top_wards":        scores.data,
        "category_breakdown": category_counts,
        "total_open":       len(grievances.data),
    }


@router.get("/wards")
def ward_scores(
    sb: Any = Depends(get_supabase),
    current: dict = Depends(get_current_user),
):
    result = sb.table("ward_sentiment_scores").select("*, wards(name, constituency)").execute()
    return result.data


@router.get("/heatmap")
def heatmap_data(
    sb: Any = Depends(get_supabase),
    current: dict = Depends(get_current_user),
):
    result = sb.table("grievances").select("location, status, ai_category, created_at").execute()
    return result.data


@router.get("/alerts")
def get_alerts(
    sb: Any = Depends(get_supabase),
    current: dict = Depends(get_current_user),
):
    now = datetime.now(timezone.utc)
    cutoff_72h = (now - timedelta(hours=72)).isoformat()

    # Critical open tickets
    critical_res = (
        sb.table("grievances")
        .select("id, tracking_id, title, ai_category, priority, status, created_at")
        .eq("priority", "critical")
        .neq("status", "resolved")
        .order("created_at", desc=True)
        .limit(5)
        .execute()
    )

    # SLA breached tickets (open > 72h)
    sla_res = (
        sb.table("grievances")
        .select("id, tracking_id, title, ai_category, priority, status, created_at")
        .neq("status", "resolved")
        .lt("created_at", cutoff_72h)
        .order("created_at", desc=False)
        .limit(5)
        .execute()
    )

    # Escalated tickets
    escalated_res = (
        sb.table("grievances")
        .select("id, tracking_id, title, ai_category, priority, status, created_at")
        .eq("status", "escalated")
        .order("created_at", desc=True)
        .limit(5)
        .execute()
    )

    alerts = []

    for g in (critical_res.data or []):
        alerts.append({
            "id": g["id"],
            "title": f"Critical: {g['title']}",
            "description": f"Tracking ID {g.get('tracking_id', '')} — {g.get('ai_category', 'General')} — awaiting resolution.",
            "severity": "critical",
            "type": "critical_grievance",
        })

    for g in (sla_res.data or []):
        hours_open = round((now - datetime.fromisoformat(g["created_at"].replace("Z", "+00:00"))).total_seconds() / 3600)
        alerts.append({
            "id": g["id"],
            "title": f"SLA Breach: {g['title']}",
            "description": f"Open for {hours_open}h — {g.get('ai_category', 'General')}. Immediate action required.",
            "severity": "high",
            "type": "sla_breach",
        })

    for g in (escalated_res.data or []):
        alerts.append({
            "id": g["id"],
            "title": f"Escalated: {g['title']}",
            "description": f"Ticket {g.get('tracking_id', '')} has been escalated — {g.get('ai_category', 'General')}.",
            "severity": "high",
            "type": "escalated",
        })

    # Deduplicate by id
    seen = set()
    unique_alerts = []
    for a in alerts:
        if a["id"] not in seen:
            seen.add(a["id"])
            unique_alerts.append(a)

    return unique_alerts


# ── Topic Clustering ────────────────────────────────────────────
@router.get("/topics")
def topic_clusters(
    sb: Any = Depends(get_supabase),
    current: dict = Depends(get_current_user),
):
    """Group open grievances by AI category for topic cluster visualization."""
    res = sb.table("grievances").select("ai_category, ai_sentiment, priority").neq("status", "resolved").execute()
    rows = res.data or []

    clusters: dict[str, dict] = {}
    for r in rows:
        cat = r.get("ai_category") or "General"
        if cat not in clusters:
            clusters[cat] = {"count": 0, "critical": 0, "negative": 0}
        clusters[cat]["count"] += 1
        if r.get("priority") == "critical":
            clusters[cat]["critical"] += 1
        if r.get("ai_sentiment") in ("negative", "very_negative"):
            clusters[cat]["negative"] += 1

    topics = sorted(
        [{"topic": k, **v} for k, v in clusters.items()],
        key=lambda x: x["count"], reverse=True,
    )
    return topics


# ── 7-Day Trend Data ────────────────────────────────────────────
@router.get("/trends")
def trends_7day(
    sb: Any = Depends(get_supabase),
    current: dict = Depends(get_current_user),
):
    """Daily grievance counts for last 7 days, broken by category."""
    now = datetime.now(timezone.utc)
    seven_ago = (now - timedelta(days=7)).isoformat()

    res = (
        sb.table("grievances")
        .select("ai_category, created_at, status")
        .gte("created_at", seven_ago)
        .execute()
    )
    rows = res.data or []

    daily: dict[str, dict] = {}
    for i in range(7):
        day = (now - timedelta(days=6 - i)).strftime("%Y-%m-%d")
        daily[day] = {"date": day, "total": 0, "resolved": 0}

    for r in rows:
        day = r["created_at"][:10]
        if day in daily:
            daily[day]["total"] += 1
            if r.get("status") == "resolved":
                daily[day]["resolved"] += 1

    return list(daily.values())


# ── Constituency Comparison ─────────────────────────────────────
@router.get("/comparison")
def constituency_comparison(
    sb: Any = Depends(get_supabase),
    current: dict = Depends(get_current_user),
):
    """Compare grievance stats across categories — resolution rate, avg sentiment."""
    res = sb.table("grievances").select("ai_category, priority, status, ai_sentiment").execute()
    rows = res.data or []

    cat_stats: dict[str, dict] = {}
    for r in rows:
        cat = r.get("ai_category") or "General"
        if cat not in cat_stats:
            cat_stats[cat] = {"total": 0, "resolved": 0, "critical": 0, "negative": 0}
        cat_stats[cat]["total"] += 1
        if r.get("status") == "resolved":
            cat_stats[cat]["resolved"] += 1
        if r.get("priority") == "critical":
            cat_stats[cat]["critical"] += 1
        if r.get("ai_sentiment") in ("negative", "very_negative"):
            cat_stats[cat]["negative"] += 1

    comparison = []
    for cat, stats in cat_stats.items():
        resolution_rate = round(stats["resolved"] / stats["total"] * 100) if stats["total"] else 0
        satisfaction = round((1 - stats["negative"] / stats["total"]) * 100) if stats["total"] else 100
        comparison.append({
            "category": cat,
            "total": stats["total"],
            "resolved": stats["resolved"],
            "resolution_rate": resolution_rate,
            "satisfaction_score": satisfaction,
            "critical_count": stats["critical"],
        })
    comparison.sort(key=lambda x: x["total"], reverse=True)
    return comparison

File: functions/sentinel.js
const { createClient } = require("@supabase/supabase-js");

exports.sentinelHandler = async (req, res) => {
  const supabaseUrl = process.env.SUPABASE_URL || '';
  const supabaseServiceRole = process.env.SUPABASE_SERVICE_ROLE_KEY || '';
  const supabase = createClient(supabaseUrl, supabaseServiceRole);

  const { action } = req.body || {};

  try {
    if (action === 'alerts') {
      const now = new Date();
      const cutoff72h = new Date(now.getTime() - 72 * 60 * 60 * 1000).toISOString();

      const [[criticalData], [slaData], [escalatedData]] = await Promise.all([
        supabase.from("grievances").select("id, tracking_id, title, ai_category, priority, status, created_at").eq("priority", "critical").neq("status", "resolved").order("created_at", { ascending: false }).limit(5),
        supabase.from("grievances").select("id, tracking_id, title, ai_category, priority, status, created_at").neq("status", "resolved").lt("created_at", cutoff72h).order("created_at", { ascending: true }).limit(5),
        supabase.from("grievances").select("id, tracking_id, title, ai_category, priority, status, created_at").eq("status", "escalated").order("created_at", { ascending: false }).limit(5)
      ].map(p => p.then(result => [result.data || [], result.error])));

      const alerts = [];
      const critical = Array.isArray(criticalData) ? criticalData : [];
      critical.forEach((g) => {
        alerts.push({
          id: g.id,
          title: `Critical: ${g.title}`,
          description: `Tracking ID ${g.tracking_id || ''} — ${g.ai_category || 'General'} — awaiting resolution.`,
          severity: "critical",
          type: "critical_grievance"
        });
      });

      const sla = Array.isArray(slaData) ? slaData : [];
      sla.forEach((g) => {
        const createdMs = Date.parse(g.created_at || now.toISOString());
        const hoursOpen = Math.round((now.getTime() - createdMs) / 3600000);
        alerts.push({
          id: g.id,
          title: `SLA Breach: ${g.title}`,
          description: `Open for ${hoursOpen}h — ${g.ai_category || 'General'}. Immediate action required.`,
          severity: "high",
          type: "sla_breach"
        });
      });

      const escalated = Array.isArray(escalatedData) ? escalatedData : [];
      escalated.forEach((g) => {
        alerts.push({
          id: g.id,
          title: `Escalated: ${g.title}`,
          description: `Ticket ${g.tracking_id || ''} has been escalated — ${g.ai_category || 'General'}.`,
          severity: "high",
          type: "escalated"
        });
      });

      const seen = new Set();
      const uniqueAlerts = alerts.filter(a => {
        if (seen.has(a.id)) return false;
        seen.add(a.id);
        return true;
      });

      return res.status(200).json(uniqueAlerts);
    }

    if (action === 'topics') {
      const { data, error } = await supabase.from("grievances").select("ai_category, ai_sentiment, priority").neq("status", "resolved");
      if (error) throw error;

      const clusters = {};
      (data || []).forEach(r => {
        const cat = r.ai_category || "General";
        if (!clusters[cat]) clusters[cat] = { count: 0, critical: 0, negative: 0 };
        clusters[cat].count += 1;
        if (r.priority === "critical") clusters[cat].critical += 1;
        if (r.ai_sentiment === "negative" || r.ai_sentiment === "very_negative") clusters[cat].negative += 1;
      });

      const topics = Object.entries(clusters).map(([topic, v]) => ({ topic, ...v })).sort((a, b) => b.count - a.count);
      return res.status(200).json(topics);
    }

    if (action === 'trends') {
      const now = new Date();
      const sevenAgo = new Date(now.getTime() - 7 * 24 * 60 * 60 * 1000).toISOString();
      const { data, error } = await supabase.from("grievances").select("ai_category, created_at, status").gte("created_at", sevenAgo);
      if (error) throw error;

      const daily = {};
      for (let i = 0; i < 7; i++) {
        const d = new Date(now.getTime() - (6 - i) * 24 * 60 * 60 * 1000);
        const day = d.toISOString().slice(0, 10);
        daily[day] = { date: day, total: 0, resolved: 0 };
      }

      (data || []).forEach(r => {
        const day = (r.created_at || '').slice(0, 10);
        if (daily[day]) {
          daily[day].total += 1;
          if (r.status === "resolved") daily[day].resolved += 1;
        }
      });

      return res.status(200).json(Object.values(daily));
    }

    if (action === 'comparison') {
      const { data, error } = await supabase.from("grievances").select("ai_category, priority, status, ai_sentiment");
      if (error) throw error;

      const catStats = {};
      (data || []).forEach(r => {
        const cat = r.ai_category || "General";
        if (!catStats[cat]) catStats[cat] = { total: 0, resolved: 0, critical: 0, negative: 0 };
        catStats[cat].total += 1;
        if (r.status === "resolved") catStats[cat].resolved += 1;
        if (r.priority === "critical") catStats[cat].critical += 1;
        if (r.ai_sentiment === "negative" || r.ai_sentiment === "very_negative") catStats[cat].negative += 1;
      });

      const comparison = Object.entries(catStats).map(([cat, stats]) => {
        const resolutionRate = stats.total ? Math.round((stats.resolved / stats.total) * 100) : 0;
        const satisfaction = stats.total ? Math.round((1 - (stats.negative / stats.total)) * 100) : 100;
        return {
          category: cat,
          total: stats.total,
          resolved: stats.resolved,
          resolution_rate: resolutionRate,
          satisfaction_score: satisfaction,
          critical_count: stats.critical
        };
      }).sort((a, b) => b.total - a.total);

      return res.status(200).json(comparison);
    }

    return res.status(400).json({ error: "Invalid action" });

  } catch (error) {
    return res.status(500).json({ error: error.message });
  }
};
